### PR TITLE
JIT: change profile slop assert to a jitdump note

### LINKED ISSUE #77450: [pgo] Assertion failed '((-slop) <= diff) && (diff <= slop)' during 'Compute edge weights (2, false)'
Pipeline: runtime-coreclr pgo
Config: win-arm64 Checked fullpgo_random_gdv_methodprofiling_only
Link: https://dev.azure.com/dnceng-public/public/_build/results?buildId=62149&view=ms.vss-test-web.build-test-results-tab&runId=1274480&resultId=108923&paneView=debug

```scala
Assert failure(PID 11476 [0x00002cd4], Thread: 7904 [0x1ee0]): Assertion failed '((-slop) <= diff) && (diff <= slop)' in 'System.RuntimeType:GetConstructorCandidates(System.String,int,int,System.Type[],bool):ListBuilder`1[System.Reflection.ConstructorInfo]:this' during 'Compute edge weights (2, false)' (IL size 103; hash 0xfc7100fa; Tier1)

File: D:\a\_work\1\s\src\coreclr\jit\fgprofile.cpp Line: 3941
Image: D:\h\w\BE4A09E7\p\corerun.exe
```

cc @AndyAyersMS

### PR DESCRIPTION
**Pull request by AndyAyersMS**
Stop asserting if we see unusually large discrepancies in the outgoing profile flow from a block. Instead just make a note in the jit dump.

Fixes #77450.

### CODE REVIEW

**Comment by AndyAyersMS** [comment]
@EgorBo PTAL
cc @dotnet/jit-contrib 

Note edge profile weights are slated to be revised during .NET 8, so this is just an interim fix to stop this assert from firing.

**Review by EgorBo — APPROVED** [review]
Wonder if we should have a special format for sort of warnings in the DUMP

**Comment by AndyAyersMS** [comment]
Failure is a crash in ILC building crossgen2. Likely unrelated since this PR is only changing an assert into a dump.
```
  crossgen2 -> /__w/1/s/artifacts/bin/coreclr/linux.arm64.Debug/crossgen2/crossgen2.dll
  Generating native code
  Segmentation fault (core dumped)
/__w/1/s/artifacts/transport/coreclr/build/Microsoft.NETCore.Native.targets(281,5): error MSB3073: The command ""/__w/1/s/artifacts/transport/coreclr/x64/ilc/ilc" @"/__w/1/s/artifacts/obj/coreclr/crossgen2/arm64/Debug/native/crossgen2.ilc.rsp"" exited with code 139. [/__w/1/s/src/coreclr/tools/aot/crossgen2/crossgen2.csproj]
##[error]artifacts/transport/coreclr/build/Microsoft.NETCore.Native.targets(281,5): error MSB3073: (NETCORE_ENGINEERING_TELEMETRY=Build) The command ""/__w/1/s/artifacts/transport/coreclr/x64/ilc/ilc" @"/__w/1/s/artifacts/obj/coreclr/crossgen2/arm64/Debug/native/crossgen2.ilc.rsp"" exited with code 139.


```

**Comment by AndyAyersMS** [comment]
Hmm, the failure is persistent. I am going to see if I can get a local repro.

@MichalStrehovsky is there any way to get crash dumps from CI when ILC fails like this? Or a stack trace?

**Comment by MichalStrehovsky** [comment]
> @MichalStrehovsky is there any way to get crash dumps from CI when ILC fails like this? Or a stack trace?

That would be a question for @dotnet/runtime-infrastructure - I don't think we are able to collect crash dumps in build legs. This is using live-built ILC with live-built JIT, hosted on top of whatever CoreCLR was used to build the repo. Such capability would have been useful in the past, and not just when using live-built tools.

I'll trigger more thorough NativeAOT testing run just in case we can hit this on a less obscure configuration.

**Comment by AndyAyersMS** [comment]
> I'll trigger more thorough NativeAOT testing run just in case we can hit this on a less obscure configuration.

Seems like lots of issues?

This change only affects jit internal diagnostics and is changing an assert into something that won't assert. So hard for me to see how it could cause any sort of failures.

**Comment by MichalStrehovsky** [comment]
> > I'll trigger more thorough NativeAOT testing run just in case we can hit this on a less obscure configuration.
> 
> Seems like lots of issues?

Yes, we had a regression (that is now fixed), but this would show up as a build break, not test break and those are easy to see. The only build break I see is in the `Build windows-x64 Release NativeAOT_Checked_Libs` leg and that one is #81460.

> This change only affects jit internal diagnostics and is changing an assert into something that won't assert. So hard for me to see how it could cause any sort of failures.

I'm a bit worried if you say this happened on repeated run. I'm not aware of seeing this in other PRs. I guess we could merge this and if this starts happening everywhere, we can start suspecting something bad like a C++ codegen bug.

**Comment by AndyAyersMS** [comment]
Let me bounce this and see if it still repros with merged-up bits.

**Comment by AndyAyersMS** [comment]
arm64 linux CI machines see to be mis-provisioned or something?
```
CMake Error at /usr/share/cmake-3.25/Modules/CMakeTestCCompiler.cmake:70 (message):
  The C compiler

    "/usr/bin/clang-9"

  is not able to compile a simple test program.
```

**Comment by AndyAyersMS** [comment]
retry didn't fix the build issues, so bounced the PR once more.

**Comment by AndyAyersMS** [comment]
ILC failure didn't repro. Extra platform failures are unrelated (and mostly fixed elsewhere).

**Comment by MichalStrehovsky** [comment]
#81476 now hit the crossgen2 musl arm64 crossbuild crash that we observed here. The run started after this PR merged to main so it's still inconclusive as to whether this change triggered it. I've not seen this failure before.

**Comment by AndyAyersMS** [comment]
Hopefully we can figure this out; let me know if I can help somehow.

**Comment by ivanpovazan** [comment]
> https://github.com/dotnet/runtime/pull/81476 now hit the crossgen2 musl arm64 crossbuild crash that we observed here. The run started after this PR merged to main so it's still inconclusive as to whether this change triggered it. I've not seen this failure before. Cc @ivanpovazan

After a rerun, the failure was not observed.

**Pull request merged**
